### ISSUE TITLE
Improve electrostatic constant treatment in ParmEd

### ISSUE THREAD

**Issue by swails** [ParmEd/ParmEd#426]
Every program uses a *slightly* different value of the electrostatic force constant that it uses.  Currently, the charge is read in unmodified from the original file where they are provided as elementary charge units, or scaled according to the Amber electrostatic constant in the case of Amber topology files.

However, when converting from Gromacs to AMBER, or AMBER to OpenMM, or AMBER to Gromacs, the charges are not scaled by the ratios of the scaling constants used in each program, which means that each program uses a *slightly* different set of charges (same net charge for a neutral system).

The way to fix this is to give ParmEd its *own* electrostatic constant (that I'll probably take directly from NIST or the CRC) that it stores charges in internally, and then have the charges get scaled appropriately according to the electrostatic constant of the parent program for the output file format.  This should ensure that the charge representations inside each program is exactly the same following conversion (subject to the limitations of floating point precision when dividing large-ish numbers that are nearly equal).

**Comment by davidlmobley** [comment]
Great!

On Tue, Sep 15, 2015 at 6:26 AM, Jason Swails <notifications@github.com>
wrote:

> Every program uses a *slightly* different value of the electrostatic
> force constant that it uses. Currently, the charge is read in unmodified
> from the original file where they are provided as elementary charge units,
> or scaled according to the Amber electrostatic constant in the case of
> Amber topology files.
>
> However, when converting from Gromacs to AMBER, or AMBER to OpenMM, or
> AMBER to Gromacs, the charges are not scaled by the ratios of the scaling
> constants used in each program, which means that each program uses a
> *slightly* different set of charges (same net charge for a neutral
> system).
>
> The way to fix this is to give ParmEd its *own* electrostatic constant
> (that I'll probably take directly from NIST or the CRC) that it stores
> charges in internally, and then have the charges get scaled appropriately
> according to the electrostatic constant of the parent program for the
> output file format. This should ensure that the charge representations
> inside each program is exactly the same following conversion (subject to
> the limitations of floating point precision when dividing large-ish numbers
> that are nearly equal).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ParmEd/ParmEd/issues/426>.
>



-- 
David Mobley
Associate Professor
Department of Pharmaceutical Sciences
Department of Chemistry
3134B Natural Sciences I
University of California, Irvine
Irvine, CA 92697
dmobley@uci.edu
work (949) 824-6383
cell (949) 385-2436

**Comment by mrshirts** [comment]
However, except for parmtop, I think most other programs store the charges as units of elementary charge.  So for those programs, I don't know that there should be a value of the Coulomb constant; they would handle that internally (and should be allowed to).

I think it comes down to what the fundamental unit is.  Since most programs that calculate charges return charges in units of elementary charge, not elementary charge times sqrt(Coulomb constant) -- even the version of RESP that antechamber uses, as far as I can tell from the code -- then ParmEd should only need that value when using a program that does not have fundamental units of charge in the input files.  Otherwise you get things like TIP3P water having a charge that is not -0.834 on the O.

**Comment by swails** [comment]
> I think it comes down to what the fundamental unit is. Since most programs that calculate charges return charges in units of elementary charge, not elementary charge times sqrt(Coulomb constant) -- even the version of RESP that antechamber uses, as far as I can tell from the code -- then ParmEd should only need that value when using a program that does not have fundamental units of charge in the input files. Otherwise you get things like TIP3P water having a charge that is not -0.834 on the O.

The problem is still there for programs that are not Amber -- it is simply masked.  When you no longer consider 1/4*pi*epsilon0 to be constant (since it differs from program to program), the differences in the electrostatic constants between programs effectively becomes part of the charges themselves.

That is, a charge of -0.834 when simulated in GROMACS is *different* than a charge of -0.834 when simulated in AMBER.  If you want to get *exactly* the same answer in both Package X and Package Y, where Package X uses electrostatic constant A and Package Y uses electrostatic constant B, and you read the charges dumped by Package X, you need to scale the charges by B/A so that the product "C*q1*q2" comes out the same in both packages.

The *only* difference here between Amber and GROMACS/CHARMM/OpenMM is that Amber hard-codes their electrostatic factor in tleap while the others hard-code theirs in the nonbonded kernel.  You still have to account for that difference if it is very important that you get identical answers in every program.  Since these constants are hard-coded in the respective packages, the only way to account for the differences is to smear it out in the charges.

That said, I don't like my approach in #427; I'm going to try something else that's a bit simpler.

**Comment by mrshirts** [comment]
This is formally all true.  However, there is a qualitative difference in
the deviations between programs. The error for AMBER is about 2 orders of
magnitude away from the DESMOND, lammps, and GROMACS numbers, which are all
within the NIST number, within uncertainty.

I'll go with the Coulomb constant in kcal/angstrom, not the square root
here.

NIST:           332.063713  +/- between 0.000044 or 0.000029, depending on
error propagation (I can share my spreadsheet on this)

Program                 value                 Stdev away.
antechamber      332.0522173   257.1582382
Gromacs            332.063693         0.447816118
Desmond           332.063762        -1.095706602
LAMMPS           332.06371            0.067527912

**Comment by swails** [comment]
This would seem to be on hold because we're waiting on whether the Amber constant will be adjusted at all (and if so, how that should be done).  ParmEd certainly *could* roll up the differences in electrostatic scaling constant into the charges themselves so that all programs gave identical answers, but it would be nice to be able to avoid that.

**Comment by swails** [comment]
Amber is the primary culprit here, since its electrostatic constant is so much different than every other program.  We should address this in Amber, not ParmEd.
